rtmr/extend/extend.go

```diff
@@ -0,0 +1,199 @@
+	indexPath := r.attribute(TsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	if err := r.ValidateRtmrIndex(); err != nil {
+		return fmt.Errorf("the tcg_map is invalid %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// ValidateRtmrIndex checks if the rtmr to PCR map is valid.
+// It returns an error if the rtmr_tcg map does not match the expected value.
+func (r *RtmrExtend) ValidateRtmrIndex() error {
+	if (r.client == nil) || (r.entry == nil) {
+		return fmt.Errorf("RtmrExtend is not initialized")
+	}
+	data, err := r.client.ReadFile(r.attribute(TsmPathTcgMap))
+	if err != nil {
+		return fmt.Errorf("could not read  %s: %v", TsmPathTcgMap, err)
+	}
+	var rtmrPcrMaps = map[int]string{
+		0: "1,7\n",
+		1: "2-6\n",
+		2: "8-15\n",
+		3: "\n",
+	}
+
+	inputMap := convertCstring(data)
+	if inputMap != rtmrPcrMaps[r.RtmrIndex] {
+		return fmt.Errorf("tcg map error. expect:%s get:%s", rtmrPcrMaps[r.RtmrIndex], string(data))
+	}
+	return nil
+}
+
+// SearchRtmrInterface searches for an rtmr entry in the configfs.
+func SearchRtmrInterface(client configfsi.Client, index int) (*RtmrExtend, error) {
+	root := TsmRtmrPrefix
+	out := RtmrExtend{}
+	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			p, err := configfsi.ParseTsmPath(path)
+			if err != nil {
+				return err
+			}
+			r := RtmrExtend{
+				RtmrIndex: index,
+				entry:     &configfsi.TsmPath{Subsystem: RtmrSubsystem, Entry: p.Entry},
+				client:    client,
+			}
+			if r.ValidateRtmrIndex() == nil {
+				out = r
+				return nil
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	if out.ValidateRtmrIndex() != nil {
+		return nil, fmt.Errorf("could not find rtmr entry in configfs: %v", err)
+	}
+
+	return &out, nil
+}
+
+// CreateRtmrInterface creates a new rtmr entry in the configfs.
+func CreateRtmrInterface(client configfsi.Client, index int) (r *RtmrExtend, err error) {
+	entryPath, err := client.MkdirTemp(TsmRtmrPrefix, fmt.Sprintf("rtmr%d-", index))
+	if err != nil {
+		return nil, fmt.Errorf("could not create rtmr entry in configfs: %v", err)
+	}
+	p, _ := configfsi.ParseTsmPath(entryPath)
+
+	r = &RtmrExtend{
+		RtmrIndex: index,
+		entry:     &configfsi.TsmPath{Subsystem: RtmrSubsystem, Entry: p.Entry},
+		client:    client,
+	}
+
+	if err := r.SetRtmrIndex(); err != nil {
+		return nil, fmt.Errorf("could not set rtmr index %d: %v", index, err)
+	}
+	return r, nil
+}
+
+// ExtendtoRtmrClient extends the measurement to the rtmr with the given client.
+func ExtendtoRtmrClient(client configfsi.Client, rtmr int, hashAlgo crypto.Hash, content []byte) error {
+	if hashAlgo != crypto.SHA384 {
+		return fmt.Errorf("unsupported hash algorithm %v", hashAlgo)
+	}
+	if len(content) == 0 {
+		return fmt.Errorf("input event log is empty")
```

Comment:
This assumes the content is an event log, which is not the case for general usage.